### ISSUE TITLE
Allow export of table data in multiple formats (XLSX, CSV, PDF)

### ISSUE THREAD

**Issue by tagwato** [dpgaspar/Flask-AppBuilder#1225]
Hi,
I found Flask-Appbuilder only 2 days ago, and I am loving it.

I'd like to request  one more new feature, which (I think) can boost users productivity:
- in list views include optional definitions to allow export of the FULL content of a table as files with these formats: Excel (XLSX), CSV and PDF.

The implementation of this feature would require F.A.B to generate some buttons or similar components to be used when a table is presented/listed to the user. When/If user presses the corresponding "button", ALL content of the table (including non visible paginated data) would be exported, with choosed format. 

The reason for this request is that, nowadays, more and more users need to do  data analysis with web contents, as well as share and store data in human readable format.
This is true for business data and, even more, for governamental data generates by official agencies.
Some client-side "tabular data" frameworks already have such important feature, for instance: 
https://datatables.net/extensions/buttons/examples/initialisation/export.html.

Thank you very much!

**Comment by apop880** [comment]
@tagwato This is already possible by exposing an endpoint on your ModelView with the needed function. Here's some sample code that will handle bulk XLSX export of your entire table.

In your imports, make sure you have the following somewhere:
```
from flask import send_file
from flask_appbuilder import expose
import pandas as pd
from io import BytesIO
import xlsxwriter
```

Here's the code that would go inside your ModelView class:
```
@expose("/export")
def excel_dl(self):
    output = BytesIO()
    
    df = pd.read_sql(sql = "select * from table_name_here",
                    con = db.session.bind)

    writer = pd.ExcelWriter(output, engine="xlsxwriter")
    df.to_excel(writer, 'data', index=False)
    writer.save()
    output.seek(0)
    return send_file(output, attachment_filename='export.xlsx', as_attachment=True)
```

From there, you'd need to customize a widget template to create a link to that endpoint (`base_url/modelviewclassname/export`)

**Comment by tagwato** [comment]
> This is already possible by exposing an endpoint on your ModelView with the needed function. Here's some sample code that will handle bulk XLSX export of your entire table...

Thanks for the suggestion, @apop880 !
I agree that there are ways to add these features to the applications, programmatically. 

Nevertheless, I believe that frameworks such as F.A.B. could easily supply the demand we see today for "no-code" or "low-code" solutions which aim to facilitate use by people who have little knowledge of programming (Yes, I dream of F.A.B being used as a replacement for MS-Excel and MS-Access "systems" – not a joke, they still exist).

The inclusion of these features directly in the framework would be an important step towards that goal. Another one would be a simple way to set masked fields in forms (see this issue: #1224).

For the record, I used to be a programmer/developer a long time ago and I can still handle 'coding'. 
But I recently experienced a situation in which the delegation of tasks to a less experienced collaborator (intern) was not possible due to the absence of these features in F.A.B.

In summary, I strongly believe that there are many people out there who could benefit if these features were natively provided by F.A.B.

**Comment by agxc** [comment]
Interesting to bump into this thread. I think this has been an re-occurring request from multiple-users in the past. I definitely agree very strongly with some of the users here that it is best we have some default options in F.A.B to allow users to export records in forms of CSV/Excel. The feature is quite common in other packages such as Flask-admin.
